src/transformers/tokenization_utils.py

```diff
@@ -49,6 +49,84 @@
 TOKENIZER_CONFIG_FILE = "tokenizer_config.json"
 
 
+class Trie:
+    """
+    Trie in Python. Creates a Trie out of a list of words. The trie is used to split on `added_tokens` in one pass
+    """
+
+    def __init__(self):
+        self.data = {}
+
+    def add(self, word):
+        if not word:
+            # Prevent empty string
+            return
+        ref = self.data
+        for char in word:
+            ref[char] = char in ref and ref[char] or {}
+            ref = ref[char]
+        ref[""] = 1
+
+    def split(self, text):
```

Comment:
(nit) Also pretty hard to understand what's happening here IMO - maybe add some more docstring as well?